ls8/cpu.c

```diff
@@ -50,9 +62,36 @@ void cpu_run(struct cpu *cpu)
   while (running) {
     // TODO
     // 1. Get the value of the current instruction (in address PC).
+    int PC = cpu->PC;
+    unsigned char instruction = cpu_ram_read(cpu, PC);
+    unsigned char operandA = cpu_ram_read(cpu, PC+1);
+    unsigned char operandB = cpu_ram_read(cpu, PC+2);
     // 2. switch() over it to decide on a course of action.
     // 3. Do whatever the instruction should do according to the spec.
+    switch(instruction) {
+      case HLT:
+        printf("HLT\n");
+        running = 0;
+        break;
+      case LDI:
+        printf("LDI\n");
+        cpu->registers[(int)operandA] = operandB;
+        cpu->PC += 3;
+        // printf("%d", cpu->registers[0]);
+        break;
+      case PRN:
+        printf("PRN\n");
+        printf("%d\n", cpu->registers[(int)operandA]);
+        cpu->PC += 2;
+        break;
+      default:
+        printf("%d", instruction);
+        printf("Default case reached.\n");
+        break;
+    }
     // 4. Move the PC to the next instruction.
+    // cpu->PC += (int)(instruction >> 6);
```

Comment:
In actuality, the code above is sound. There's one thing: if you notice the workaround code above, notice that we move the `PC` 2 for `PRN`, even though there 's only one operand after PRN, and 3 for `LDI`, despite being two operands. The reason we're moving it one more is that the PC needs to be pointing to the next instruction to be executed after the PC moves, but if you just move it by the number of operands it has, you'll find yourself in the wrong place, often the operand before the next instruction.

TL;DR 
```suggestion
    cpu->PC += (int)(instruction >> 6) + 1;
```

And remove the hard-coded PC incrementors above.

Reply:
I just put this back in before lecture today with the +1.